collector/zfs_freebsd.go

```diff
@@ -0,0 +1,115 @@
+package collector
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+/*
+#cgo LDFLAGS:
+#include <sys/param.h>
+#include <sys/module.h>
+
+int zfsModuleLoaded() {
+	int modid = modfind("zfs");
+	return modid < 0 ? 0 : -1;
+}
+
+*/
+import "C"
+
+func (c *zfsCollector) zfsAvailable() error {
+	if C.zfsModuleLoaded() == 0 {
+		return zfsNotAvailableError
+	}
+	return nil
+}
+
+const zfsArcstatsSysctl = "kstat.zfs.misc.arcstats"
+
+func (c *zfsCollector) RunOnStdout(cmd *exec.Cmd, handler func(io.Reader) error) (err error) {
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return
+	}
+
+	if err = cmd.Start(); err != nil {
+		return
+	}
+
+	err = handler(stdout)
+	if err != nil {
+		return
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return
+	}
+
+	return err
+
+}
+
+func (c *zfsCollector) updateArcstats(ch chan<- prometheus.Metric) (err error) {
+
+	cmd := exec.Command("sysctl", zfsArcstatsSysctl)
```

Comment:
Yes, we don't allow shelling out in the node exporter. You should either make the syscalls directly, or use a cronjob and the textfile collector.